File: python-algorithm/algo_03_two_points/L88.py
#--coding:utf-8--
'''
@Time   : 2020/9/24
@Author : Heng Li
@Email  : liheng_lulia@163.com
'''

class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        i, j, k = m - 1, n - 1, m + n - 1
        while i >= 0 and j >= 0:
            if nums1[i] > nums2[j]:
                nums1[k] = nums1[i]
                k -= 1
                i -= 1
            else:
                nums1[k] = nums2[j]
                k -= 1
                j -= 1
        # 容易漏掉，没处理完的部分
        while j >= 0:
            nums1[k] = nums2[j]
            k -= 1
            j -= 1

File: python-algorithm/algo_03_two_points/L88.cpp

//88 合并两个有序数组

//思路比较独特，倒着处理数组

//此题输入中设定数组1后面存在足够空间存放合并后的数据;要求不开辟多余的内存完成合并


//输入:
//nums1 = [1,2,3,0,0,0], m = 3
//nums2 = [2,5,6],       n = 3


#include <iostream>
#include <vector>
using namespace std;

class Solution {

public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        int i = m-1;
        int j = n-1;
        int k = m+n-1;
        while(i >= 0 && j >= 0){
            if(nums1[i] > nums2[j]){
                nums1[k--] = nums1[i];
                --i;
            }
            else{
                nums1[k--]=nums2[j];
                --j;
            }
        }
        while(j>=0)
            nums1[k--] = nums2[j--];
    }

    //合并2个有序数组到一个新的数组中
    vector<int> mergeTwoArrayList(vector<int>& nums1, int m , vector<int>& nums2 , int n ){
        vector<int> res ;
        int i = 0 , j = 0 ;
        while(i < m && j < n){
            if(nums1[i] > nums2[j]){
                res.push_back(nums2[j]);
                j++ ;
            }
            else {
                res.push_back(nums1[i]);
                i++ ;
            }
        }

        while(i < m){
            res.push_back(nums1[i]);
            i++ ;
        }

        while(j < n){
            res.push_back(nums2[j]);
            j++ ;
        }

        return res ;
    }

};